neqo-transport/src/connection.rs

```diff
@@ -5173,4 +5183,141 @@ mod tests {
+
+        let mut unknown_version_packet = vec![0x80, 0x1a, 0x1a, 0x1a, 0x1a];
+        unknown_version_packet.resize(1200, 0x0);
+
+        assert_eq!(
+            server.process(
+                Some(Datagram::new(
+                    loopback(),
+                    loopback(),
+                    unknown_version_packet,
+                )),
+                now(),
+            ),
+            Output::None
+        );
+
+        assert_eq!(1, server.stats().dropped_rx);
+    }
+
+    fn get_vn(client: &mut Connection, versions: &[u32]) -> Vec<u8> {
+        // Start the handshake.
+        let out = client
+            .process(None, now())
+            .dgram()
+            .expect("a datagram")
+            .to_vec();
+
+        let mut dec = Decoder::from(&out[5..]); // Skip past version.
+        let dcid = dec.decode_vec(1).expect("client DCID");
+        let scid = dec.decode_vec(1).expect("client SCID");
+
+        let mut encoder = Encoder::default();
+        encoder.encode_byte(PACKET_BIT_LONG);
+        encoder.encode(&[0; 4]); // Zero version == VN.
+        encoder.encode_vec(1, dcid);
+        encoder.encode_vec(1, scid);
+
+        for v in versions {
+            encoder.encode_uint(4, *v);
+        }
+        encoder.into()
+    }
+
+    #[test]
+    fn version_negotiation_current_version() {
+        let mut client = default_client();
+
+        let vn = get_vn(&mut client, &[0x1a1_a1a1a, QUIC_VERSION]);
+
+        assert_eq!(
+            client.process(Some(Datagram::new(loopback(), loopback(), vn,)), now(),),
+            Output::Callback(Duration::from_millis(120))
+        );
+        assert_eq!(*client.state(), State::WaitInitial);
+        assert_eq!(1, client.stats().dropped_rx);
+    }
+
+    #[test]
+    fn version_negotiation_only_reserved() {
+        let mut client = default_client();
+
+        let vn = get_vn(&mut client, &[0x1a1a_1a1a, 0x2a2a_2a2a]);
+
+        assert_eq!(
+            client.process(Some(Datagram::new(loopback(), loopback(), vn,)), now(),),
+            Output::None
+        );
+        match client.state() {
+            State::Closed(err) => {
+                assert_eq!(*err, ConnectionError::Transport(Error::VersionNegotiation))
+            }
+            _ => panic!("Invalid client state"),
+        }
+    }
+
+    #[test]
+    fn version_negotiation_corrupted() {
+        let mut client = default_client();
+
+        let vn = get_vn(&mut client, &[0x1a1a_1a1a, 0x2a2a_2a2a]);
+
+        assert_eq!(
+            client.process(
+                Some(Datagram::new(loopback(), loopback(), &vn[..vn.len() - 1])),
+                now(),
+            ),
+            Output::None
+        );
+        match client.state() {
+            State::Closed(err) => {
+                assert_eq!(*err, ConnectionError::Transport(Error::VersionNegotiation))
+            }
+            _ => panic!("Invalid client state"),
+        }
+    }
+
+    #[test]
+    fn version_negotiation_empty() {
```

Comment:
Shouldn't this have the same effect as the corrupted one?  That is, shouldn't we ignore it?